hazelcast/src/main/java/com/hazelcast/cp/internal/datastructures/spi/blocking/BlockingResource.java

```diff
@@ -148,7 +148,17 @@ final void closeSession(long sessionId, List<Long> expiredWaitKeys, Map<Long, Ob
 
     final void collectAttachedSessions(Collection<Long> sessions) {
         sessions.addAll(getActivelyAttachedSessions());
-        for (WaitKeyContainer<W> key : waitKeys.values()) {
```

Comment:
Couldn't we do:
```java
        for (WaitKeyContainer<W> key : new ArrayList<>(waitKeys.values())) {
            sessions.add(key.sessionId());
        }
```

Reply:
Unfortunately no, since the `ArrayList` constructor uses `toArray` https://github.com/AdoptOpenJDK/openjdk-jdk14u/blob/4c6b476f2be142954d9823cc8a2e55a95d7433f4/src/java.base/share/classes/java/util/ArrayList.java#L181
And `toArray` for `LinkedHashMap` has been implemented via Iterator before JDK 14.

The closest solution I've found is https://www.javaspecialists.eu/archive/Issue296-Concurrent-LinkedHashSet.html, but maybe it will be better to introduce some locks. 

Thanks, Łukasz for the comment!

Reply:
No concurrent collection will solve this problem for the same reason that Collections.synchronizedMap doesn't solve it: we have the problem that iterators are exposed which are then used to modify waitKeys, potentially while other threads are iterating over it.

Note: ConcurrentModificationException, which is in this case just a symptom, has per se nothing to do with multiple threads. It just means a collection is modified while being iterated over. It can be thrown by a single-threaded program. E.g. running the following code will produce it:

```java
List<Integer> values = new LinkedList<>();

values.add(0);

Iterator<Integer> it = values.iterator();

while (it.hasNext()){
    values.remove(0);
    it.next();
}
```